src/decoder.c

```diff
@@ -0,0 +1,393 @@
+                               p11prov_decoder_ctx_store_obj, ctx);
+        if (ret != CKR_OK) {
+            P11PROV_debug(
+                "Failed to find object on (slotid=%lx, session=%lx, ret=%lx)",
+                slotid, session, ret);
+            slotid = nextid;
+            continue;
+        }
+        slotid = nextid;
+    } while (nextid != CK_UNAVAILABLE_INFORMATION);
+    ret = CKR_OK;
+    p11prov_pop_error_to_mark(ctx->provctx);
+    p11prov_clear_last_error_mark(ctx->provctx);
+
+    if (ctx->invalid) {
+        ret = CKR_GENERAL_ERROR;
+        P11PROV_debug("Invalid context status");
+        goto done;
+    }
+
+    if (!ctx->object) {
+        ret = CKR_GENERAL_ERROR;
+        P11PROV_debug("No matching object stored");
+        goto done;
+    }
+
+done:
+    p11prov_uri_free(parsed_uri);
+    p11prov_return_session(session);
+    if (ret != CKR_OK) {
+        decoder_ctx_object_free(ctx);
+    }
+
+    P11PROV_debug("Done (result:%d)", ret);
+    return ret;
+}
+
+static int obj_desc_verify(P11PROV_PK11_URI *obj)
+{
+    const char *desc = NULL;
+    int desc_len;
+    desc = (const char *)ASN1_STRING_get0_data(obj->desc);
+    desc_len = ASN1_STRING_length(obj->desc);
+    if (!desc || desc_len <= 0) {
+        P11PROV_debug("Failed to get description");
+        return RET_OSSL_ERR;
+    }
+
+    if (desc_len != (sizeof(P11PROV_DESCS_URI_FILE) - 1)
+        || 0 != strncmp(desc, P11PROV_DESCS_URI_FILE, desc_len)) {
+        P11PROV_debug("Description string does not match");
+        return RET_OSSL_ERR;
+    }
+    return RET_OSSL_OK;
+}
+
+static int p11prov_der_decoder_p11prov_obj_decode(
+    const char *desired_data_type, void *inctx, OSSL_CORE_BIO *cin,
+    int selection, OSSL_CALLBACK *object_cb, void *object_cbarg,
+    OSSL_PASSPHRASE_CALLBACK *pw_cb, void *pw_cbarg)
+{
+    P11PROV_DECODER_CTX *const ctx = inctx;
+    P11PROV_PK11_URI *obj = NULL;
+    BIO *bin;
+    int ret = RET_OSSL_CARRY_ON_DECODING;
+    const char *uri = NULL;
+    int uri_len;
+
+    bin = BIO_new_from_core_bio(p11prov_ctx_get_libctx(ctx->provctx), cin);
+    if (!bin) {
+        P11PROV_debug("P11 DECODER BIO_new_from_core_bio failed");
+        goto done;
+    }
+
+    const unsigned char *der;
+    long der_len = BIO_get_mem_data(bin, &der);
+    if (der_len <= 0) {
+        P11PROV_debug("P11 DECODER BIO_get_mem_data failed");
+        goto done;
+    }
+    obj = d2i_P11PROV_PK11_URI(NULL, &der, der_len);
+    if (!obj) {
+        P11PROV_debug("P11 KEY DECODER d2i_P11PROV_PK11_URI failed");
+        goto done;
+    }
+
+    if (!obj_desc_verify(obj)) {
+        goto done;
+    }
+
+    uri = (const char *)ASN1_STRING_get0_data(obj->uri);
+    uri_len = ASN1_STRING_length(obj->uri);
+    /* todo check string ends in \0 */
+    if (!uri || uri_len <= 0) {
+        P11PROV_debug("Failed to get URI");
+        goto done;
+    }
+
+    if (p11prov_decoder_load_obj(ctx, uri, pw_cb, pw_cbarg) != CKR_OK) {
```

Comment:
Actually if you call the openssl store API here that wouldn't that also remove the need for p11prov_obj_to_ossl_obj() as well?